Area Setting Arrows not rendering

I selected the area plot to display surface velocity and tried to add the velocity arrows. The arrows are not rendering.

![image](https://user-images.githubusercontent.com/30048431/78786425-4eec3100-7983-11ea-9a39-f01c12d6b649.png)


Url:
http://navigator.oceansdata.ca/public/?query=%7B%22area%22%3A%5B%7B%22innerrings%22%3A%5B%5D%2C%22name%22%3A%22%22%2C%22polygons%22%3A%5B%5B%5B51.835777520452496%2C-54.5443707704544%5D%2C%5B45.460130637921%2C-41.82952523231505%5D%2C%5B41.200699508460985%2C-46.80999577045439%5D%2C%5B41.98943703321646%2C-55.188903808593764%5D%2C%5B51.835777520452496%2C-54.5443707704544%5D%5D%5D%7D%5D%2C%22center%22%3A%5B%22-49.7070%22%2C%2253.0000%22%5D%2C%22dataset%22%3A%22cmems_seasonal_climatolog%22%2C%22depth%22%3A0%2C%22modal%22%3A%22area%22%2C%22names%22%3A%5B%5D%2C%22projection%22%3A%22EPSG%3A3857%22%2C%22showModal%22%3Atrue%2C%22subquery%22%3A%7B%22bathymetry%22%3Atrue%2C%22colormap_diff%22%3A%22default%22%2C%22contour%22%3A%7B%22colormap%22%3A%22default%22%2C%22hatch%22%3Afalse%2C%22legend%22%3Atrue%2C%22levels%22%3A%22auto%22%2C%22variable%22%3A%22none%22%7D%2C%22convertToUserGrid%22%3Afalse%2C%22currentTab%22%3A1%2C%22dataset_0%22%3A%7B%22dataset%22%3A%22cmems_seasonal_climatolog%22%2C%22dataset_attribution%22%3A%22E.U.%20Copernicus%20Marine%20Service%20Information%20(CMEMS)%22%2C%22dataset_quantum%22%3A%22month%22%2C%22depth%22%3A0%2C%22time%22%3A15672%2C%22variable%22%3A%22magwatervel%22%2C%22variable_scale%22%3A%5B0%2C1%5D%7D%2C%22dpi%22%3A144%2C%22leftColormap%22%3A%22default%22%2C%22linearthresh%22%3A200%2C%22output_endtime%22%3A15672%2C%22output_format%22%3A%22NETCDF4%22%2C%22output_starttime%22%3A15672%2C%22output_timerange%22%3Afalse%2C%22output_variables%22%3A%22%22%2C%22quiver%22%3A%7B%22colormap%22%3A%22default%22%2C%22magnitude%22%3A%22length%22%2C%22variable%22%3A%22magwatervel%22%7D%2C%22rightColormap%22%3A%22default%22%2C%22scale%22%3A%220%2C1%2Cauto%22%2C%22scale_1%22%3A%22-5%2C30%2Cauto%22%2C%22scale_diff%22%3A%22-10%2C10%2Cauto%22%2C%22setDefaultScale%22%3Atrue%2C%22showarea%22%3Afalse%2C%22size%22%3A%2210x7%22%2C%22surfacevariable%22%3A%22none%22%2C%22syncLocalToGlobalState%22%3Afalse%2C%22zip%22%3Afalse%7D%2C%22time%22%3A15672%2C%22variable%22%3A%22magwatervel%22%2C%22vectorid%22%3Anull%2C%22vectortype%22%3Anull%2C%22zoom%22%3A4%7D

Stack trace:
```sh
[2020-04-09 16:22:45 +0000] [4339] [ERROR] Error handling request /api/v1.0/plot/?query=%7B%22area%22%3A%5B%7B%22innerrings%22%3A%5B%5D%2C%22name%22%3A%22%22%2C%22polygons%22%3A%5B%5B%5B49.95121990866204%2C-42.93457031249999%5D%2C%5B50.56928286558244%2C-52.1630859375%5D%2C%5B44.59046718130884%2C-53.745117187499986%5D%2C%5B43.96119063892027%2C-42.758789062499986%5D%2C%5B49.95121990866204%2C-42.93457031249999%5D%5D%5D%7D%5D%2C%22bathymetry%22%3Atrue%2C%22colormap%22%3A%22default%22%2C%22contour%22%3A%7B%22colormap%22%3A%22default%22%2C%22hatch%22%3Afalse%2C%22legend%22%3Atrue%2C%22levels%22%3A%22auto%22%2C%22variable%22%3A%22none%22%7D%2C%22dataset%22%3A%22giops_day%22%2C%22depth%22%3A0%2C%22interp%22%3A%22gaussian%22%2C%22neighbours%22%3A10%2C%22projection%22%3A%22EPSG%3A3857%22%2C%22quantum%22%3A%22day%22%2C%22quiver%22%3A%7B%22colormap%22%3A%22default%22%2C%22magnitude%22%3A%22length%22%2C%22variable%22%3A%22magwatervel%22%7D%2C%22radius%22%3A25%2C%22scale%22%3A%22-5%2C30%2Cauto%22%2C%22showarea%22%3Atrue%2C%22time%22%3A2217585600%2C%22type%22%3A%22map%22%2C%22variable%22%3A%22votemper%22%7D&format=json
Traceback (most recent call last):
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 135, in handle
    self.handle_request(listener, req, client, addr)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 176, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/buildadm/Ocean-Data-Map-Project/routes/api_v1_0.py", line 554, in plot_v1_0
    img, mime, filename = plotter.run()
  File "/home/buildadm/Ocean-Data-Map-Project/plotting/plotter.py", line 58, in run
    _ = self.prepare_plot()
  File "/home/buildadm/Ocean-Data-Map-Project/plotting/plotter.py", line 51, in prepare_plot
    self.load_data()
  File "/home/buildadm/Ocean-Data-Map-Project/plotting/map.py", line 291, in load_data
    self.neighbours,
  File "/home/buildadm/Ocean-Data-Map-Project/data/model.py", line 122, in get_area
    latitude, longitude, depth, variable, time, return_depth=return_depth
  File "/home/buildadm/Ocean-Data-Map-Project/data/mercator.py", line 263, in get_point
    data = var[time_slice, int(depth), miny:maxy, minx:maxx]
  File "/home/buildadm/Ocean-Data-Map-Project/data/calculated.py", line 128, in __getitem__
    self._expression, self._parent, key, self._dims)
  File "/home/buildadm/Ocean-Data-Map-Project/data/calculated_parser/parser.py", line 48, in parse
    self.parser.parse(expression)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/ply/yacc.py", line 1120, in parseopt_notrack
    p.callable(pslice)
  File "/home/buildadm/Ocean-Data-Map-Project/data/calculated_parser/parser.py", line 101, in p_expression_variable
    t[0] = self.data.variables[t[1]][
  File "/home/buildadm/miniconda/3/amd64/envs/navigator/lib/python3.6/site-packages/xarray/core/utils.py", line 317, in __getitem__
    return self.mapping[key]
KeyError: 'vozocrtx'
```
```

--- a/plotting/map.py
+++ b/plotting/map.py
@@ -25,6 +25,7 @@ import plotting.colormap as colormap
 import plotting.overlays as overlays
 import plotting.utils as utils
 from data import open_dataset
+from data.utils import get_data_vars_from_equation
 from oceannavigator import DatasetConfig
 from plotting.plotter import Plotter
 from utils.errors import ClientError, ServerError
@@ -36,10 +37,13 @@ class MapPlotter(Plotter):
     def __init__(self, dataset_name: str, query: str, **kwargs):
         self.plottype: str = 'map'
 
-        super(MapPlotter, self).__init__(dataset_name, query, **kwargs)
+        super().__init__(dataset_name, query, **kwargs)
 
     def parse_query(self, query):
-        super(MapPlotter, self).parse_query(query)
+        super().parse_query(query)
+
+        if len(self.variables) > 1:
+            raise ClientError(f"MapPlotter only supports 1 variable. Received multiple: {self.variables}")
 
         self.projection = query.get('projection')
 
@@ -96,6 +100,11 @@ class MapPlotter(Plotter):
         self.quiver = query.get('quiver')
 
         self.contour = query.get('contour')
+
+    def __load_quiver(self) -> bool:
+        return self.quiver is not None and \
+               self.quiver['variable'] and \
+               self.quiver['variable'] != 'none'
 
     def load_data(self):
         distance = VincentyDistance()
@@ -184,36 +193,24 @@ class MapPlotter(Plotter):
             gridx = int(500 / self.basemap.aspect)
 
         self.longitude, self.latitude = self.basemap.makegrid(gridx, gridy)
+        variables_to_load = self.variables[:] # we don't want to change self,variables so copy it
+        if self.__load_quiver():
+            variables_to_load.append(self.quiver['variable'])
 
-        with open_dataset(self.dataset_config, variable=self.variables, timestamp=self.time) as dataset:
+        with open_dataset(self.dataset_config, variable=variables_to_load, timestamp=self.time) as dataset:
 
-            if len(self.variables) > 1:
-                self.variable_unit = self.get_vector_variable_unit(
-                    dataset, self.variables
-                )
-                self.variable_name = self.get_vector_variable_name(
-                    dataset, self.variables
-                )
-                scale_factor = self.get_vector_variable_scale_factor(
-                    dataset, self.variables
-                )
-            else:
-                self.variable_unit = self.get_variable_units(
-                    dataset, self.variables
-                )[0]
-                self.variable_name = self.get_variable_names(
-                    dataset,
-                    self.variables
-                )[0]
-                scale_factor = self.get_variable_scale_factors(
-                    dataset, self.variables
-                )[0]
+            self.variable_unit = self.get_variable_units(
+                dataset, self.variables
+            )[0]
+            self.variable_name = self.get_variable_names(
+                dataset, self.variables
+            )[0]
+            scale_factor = self.get_variable_scale_factors(
+                dataset, self.variables
+            )[0]
 
             if self.cmap is None:
-                if len(self.variables) == 1:
-                    self.cmap = colormap.find_colormap(self.variable_name)
-                else:
-                    self.cmap = colormap.colormaps.get('speed')
+                self.cmap = colormap.find_colormap(self.variable_name)
 
             if self.depth == 'bottom':
                 depth_value_map = 'Bottom'
@@ -224,31 +221,28 @@ class MapPlotter(Plotter):
                 depth_value_map = depth_value
 
             data = []
-            allvars = []
-            for v in self.variables:
-                var = dataset.variables[v]
-                allvars.append(v)
-                if self.filetype in ['csv', 'odv', 'txt']:
-                    d, depth_value_map = dataset.get_area(
-                        np.array([self.latitude, self.longitude]),
-                        self.depth,
-                        self.time,
-                        v,
-                        self.interp,
-                        self.radius,
-                        self.neighbours,
-                        return_depth=True
-                    )
-                else:
-                    d = dataset.get_area(
-                        np.array([self.latitude, self.longitude]),
-                        self.depth,
-                        self.time,
-                        v,
-                        self.interp,
-                        self.radius,
-                        self.neighbours
-                    )
+            var = dataset.variables[self.variables[0]]
+            if self.filetype in ['csv', 'odv', 'txt']:
+                d, depth_value_map = dataset.get_area(
+                    np.array([self.latitude, self.longitude]),
+                    self.depth,
+                    self.time,
+                    self.variables[0],
+                    self.interp,
+                    self.radius,
+                    self.neighbours,
+                    return_depth=True
+                )
+            else:
+                d = dataset.get_area(
+                    np.array([self.latitude, self.longitude]),
+                    self.depth,
+                    self.time,
+                    self.variables[0],
+                    self.interp,
+                    self.radius,
+                    self.neighbours
+                )
 
                 d = np.multiply(d, scale_factor)
 
@@ -262,9 +256,6 @@ class MapPlotter(Plotter):
                         self.depth_label = " at " + \
                             str(int(np.round(depth_value_map))) + " m"
 
-            if len(data) == 2:
-                data[0] = np.sqrt(data[0] ** 2 + data[1] ** 2)
-
             self.data = data[0]
 
             quiver_data = []
@@ -272,48 +263,85 @@ class MapPlotter(Plotter):
             # will only be used for CSV export.
             quiver_data_fullgrid = []
 
-            if self.quiver is not None and \
-                self.quiver['variable'] != '' and \
-                    self.quiver['variable'] != 'none':
-                for v in self.quiver['variable'].split(','):
-                    allvars.append(v)
-                    var = dataset.variables[v]
-                    quiver_unit = self.dataset_config.variable[var].unit
-                    quiver_name = self.dataset_config.variable[var].name
-                    quiver_lon, quiver_lat = self.basemap.makegrid(50, 50)
-                    d = dataset.get_area(
-                        np.array([quiver_lat, quiver_lon]),
-                        self.depth,
-                        self.time,
-                        v,
-                        self.interp,
-                        self.radius,
-                        self.neighbours,
-                    )
-                    quiver_data.append(d)
-                    # Get the quiver data on the same grid as the main
-                    # variable.
-                    d = dataset.get_area(
-                        np.array([self.latitude, self.longitude]),
-                        self.depth,
-                        self.time,
-                        v,
-                        self.interp,
-                        self.radius,
-                        self.neighbours,
-                    )
-                    quiver_data_fullgrid.append(d)
-
-                self.quiver_name = self.get_vector_variable_name(
-                    dataset, self.quiver['variable'].split(',')
+            if self.__load_quiver():
+                var = dataset.variables[self.quiver['variable']]
+                quiver_unit = self.dataset_config.variable[var].unit
+                quiver_name = self.dataset_config.variable[var].name
+                quiver_lon, quiver_lat = self.basemap.makegrid(50, 50)
+                
+                data_vars = list(set(self.dataset_config.variables) - set(self.dataset_config.calculated_variables))
+                component_vars = get_data_vars_from_equation(self.dataset_config.calculated_variables[var.key]['equation'],
+                                                            data_vars)
+                
+                # Expect component vars to be [x_variable, y_variable]
+                x_vals = dataset.get_area(
+                    np.array([quiver_lat, quiver_lon]),
+                    self.depth,
+                    self.time,
+                    component_vars[0],
+                    self.interp,
+                    self.radius,
+                    self.neighbours,
                 )
+                quiver_data.append(x_vals)
+
+                y_vals = dataset.get_area(
+                    np.array([quiver_lat, quiver_lon]),
+                    self.depth,
+                    self.time,
+                    component_vars[1],
+                    self.interp,
+                    self.radius,
+                    self.neighbours,
+                )
+                quiver_data.append(y_vals)
+                
+                mag_data = dataset.get_area(
+                    np.array([quiver_lat, quiver_lon]),
+                    self.depth,
+                    self.time,
+                    self.quiver['variable'],
+                    self.interp,
+                    self.radius,
+                    self.neighbours,
+                )
+                self.quiver_magnitude = mag_data
+
+                # Get the quiver data on the same grid as the main
+                # variable.
+                x_vals = dataset.get_area(
+                    np.array([self.latitude, self.longitude]),
+                    self.depth,
+                    self.time,
+                    component_vars[0],
+                    self.interp,
+                    self.radius,
+                    self.neighbours,
+                )
+                quiver_data_fullgrid.append(x_vals)
+
+                y_vals = dataset.get_area(
+                    np.array([self.latitude, self.longitude]),
+                    self.depth,
+                    self.time,
+                    component_vars[1],
+                    self.interp,
+                    self.radius,
+                    self.neighbours,
+                )
+                quiver_data_fullgrid.append(y_vals)
+
+                self.quiver_name = self.get_variable_names(
+                    dataset,
+                    [self.quiver['variable']]
+                )[0]
                 self.quiver_longitude = quiver_lon
                 self.quiver_latitude = quiver_lat
                 self.quiver_unit = quiver_unit
             self.quiver_data = quiver_data
             self.quiver_data_fullgrid = quiver_data_fullgrid
 
-            if all([len(dataset.variables[v].dimensions) == 3 for v in allvars]):
+            if all([dataset.variables[v].is_surface_only() for v in variables_to_load]):
                 self.depth = 0
 
             contour_data = []
@@ -360,13 +388,10 @@ class MapPlotter(Plotter):
                     )
                     data.append(d)
 
-                if len(data) == 2:
-                    data = np.sqrt(data[0] ** 2 + data[1] ** 2)
-                else:
-                    data = data[0]
+                
+                data = data[0]
 
                 self.data -= data
-
         # Load bathymetry data
         self.bathymetry = overlays.bathymetry(
             self.basemap,
@@ -376,7 +401,7 @@ class MapPlotter(Plotter):
         )
 
         if self.depth != 'bottom' and self.depth != 0:
-            if len(quiver_data) > 0:
+            if quiver_data:
                 quiver_bathymetry = overlays.bathymetry(
                     self.basemap, quiver_lat, quiver_lon)
 
@@ -459,12 +484,7 @@ class MapPlotter(Plotter):
     def csv(self):
         # If the user has selected the display of quiver data in the browser,
         # then also export that data in the CSV file.
-        if self.quiver is not None and \
-            self.quiver['variable'] != '' and \
-                self.quiver['variable'] != 'none':
-            have_quiver = True
-        else:
-            have_quiver = False
+        have_quiver = self.__load_quiver()
 
         header = [
             ['Dataset', self.dataset_name],
@@ -583,7 +603,7 @@ class MapPlotter(Plotter):
             vmax = self.scale[1]
         else:
             vmin, vmax = utils.normalize_scale(self.data,
-                                               self.dataset_config.variable[",".join(self.variables)])
+                                               self.dataset_config.variable[f"{self.variables[0]}"])
 
         c = self.basemap.imshow(
             self.data, vmin=vmin, vmax=vmax, cmap=self.cmap)
@@ -596,11 +616,10 @@ class MapPlotter(Plotter):
                                                       returnxy=True)
             qx = np.ma.masked_where(np.ma.getmask(self.quiver_data[0]), qx)
             qy = np.ma.masked_where(np.ma.getmask(self.quiver_data[1]), qy)
-            quiver_mag = np.sqrt(qx ** 2 + qy ** 2)
 
             if self.quiver['magnitude'] != 'length':
-                qx = qx / quiver_mag
-                qy = qy / quiver_mag
+                qx = qx / self.quiver_magnitude
+                qy = qy / self.quiver_magnitude
                 qscale = 50
             else:
                 qscale = None
@@ -614,7 +633,7 @@ class MapPlotter(Plotter):
                 q = self.basemap.quiver(
                     x, y,
                     qx, qy,
-                    quiver_mag,
+                    self.quiver_magnitude,
                     width=0.0035,
                     headaxislength=4, headlength=4,
                     scale=qscale,
@@ -632,7 +651,7 @@ class MapPlotter(Plotter):
                 )
 
             if self.quiver['magnitude'] == 'length':
-                unit_length = np.mean(quiver_mag) * 2
+                unit_length = np.mean(self.quiver_magnitude) * 2
                 unit_length = np.round(unit_length,
                                        -int(np.floor(np.log10(unit_length))))
                 if unit_length >= 1:

--- a/plotting/plotter.py
+++ b/plotting/plotter.py
@@ -3,6 +3,7 @@ import datetime
 import re
 from abc import ABCMeta, abstractmethod
 from io import BytesIO, StringIO
+from typing import List
 
 import matplotlib.pyplot as plt
 import numpy as np
@@ -333,7 +334,7 @@ class Plotter(metaclass=ABCMeta):
 
             return (buf.getvalue(), self.mime, self.filename)
 
-    def get_variable_names(self, dataset, variables):
+    def get_variable_names(self, dataset, variables: List[str]) -> List[str]:
         """Returns a list of names for the variables.
 
         Parameters:

